File: skills/cp-solver/references/implementations/fenwick-tree/fenwick_tree.py
"""
Fenwick Tree (Binary Indexed Tree) — Point Update, Prefix/Range Query
Complexity: O(N) build, O(log N) update/query
Chỉ hỗ trợ phép toán có nghịch đảo: sum, XOR

Usage:
    ft = FenwickTree(n)          # zero-initialized
    ft = FenwickTree(n, a)       # build from list a (0-indexed)
    ft.update(i, delta)          # a[i] += delta (0-indexed)
    ft.query(i)                  # prefix sum [0..i]
    ft.range_query(l, r)         # range sum [l..r] (0-indexed, inclusive)
"""
import sys
input = sys.stdin.readline


class FenwickTree:
    def __init__(self, n, a=None):
        self.n = n
        self.tree = [0] * (n + 1)
        if a:
            for i, v in enumerate(a):
                self.update(i, v)

    def update(self, i, delta):     # 0-indexed
        i += 1
        while i <= self.n:
            self.tree[i] += delta
            i += i & (-i)

    def query(self, i):             # prefix sum [0..i], 0-indexed
        i += 1
        s = 0
        while i > 0:
            s += self.tree[i]
            i -= i & (-i)
        return s

    def range_query(self, l, r):    # [l..r], 0-indexed, inclusive
        if l == 0:
            return self.query(r)
        return self.query(r) - self.query(l - 1)


class FenwickTree2D:
    """Fenwick Tree 2D — Point update, prefix/range sum queries on matrix."""
    def __init__(self, n, m):
        self.n = n
        self.m = m
        self.tree = [[0] * (m + 1) for _ in range(n + 1)]

    def update(self, x, y, delta):   # 0-indexed
        x += 1
        while x <= self.n:
            y1 = y + 1
            while y1 <= self.m:
                self.tree[x][y1] += delta
                y1 += y1 & (-y1)
            x += x & (-x)

    def query(self, x, y):           # prefix sum [0..x][0..y]
        x += 1
        s = 0
        while x > 0:
            y1 = y + 1
            while y1 > 0:
                s += self.tree[x][y1]
                y1 -= y1 & (-y1)
            x -= x & (-x)
        return s

    def range_query(self, x1, y1, x2, y2):  # [x1..x2][y1..y2], 0-indexed
        ans = self.query(x2, y2)
        if x1 > 0: ans -= self.query(x1-1, y2)
        if y1 > 0: ans -= self.query(x2, y1-1)
        if x1 > 0 and y1 > 0: ans += self.query(x1-1, y1-1)
        return ans


# ===== USAGE EXAMPLE =====
def solve():
    n, q = map(int, input().split())
    a = list(map(int, input().split()))
    ft = FenwickTree(n, a)

    out = []
    for _ in range(q):
        t, *args = map(int, input().split())
        if t == 1:
            i, val = args
            ft.update(i-1, val)           # 1-indexed
        else:
            l, r = args
            out.append(ft.range_query(l-1, r-1))  # 1-indexed

    sys.stdout.write('\n'.join(map(str, out)) + '\n')


if __name__ == '__main__':
    import sys
    input = sys.stdin.readline
    solve()

File: skills/cp-solver/references/implementations/fenwick-tree/fenwick_tree.cpp
// Fenwick Tree (Binary Indexed Tree) — Point Update, Prefix/Range Query
// Complexity: O(N) build, O(log N) update/query
// Chỉ hỗ trợ các phép toán có nghịch đảo: sum, XOR

#include <bits/stdc++.h>
using namespace std;

struct FenwickTree {
  int n;
  vector<long long> tree;

  FenwickTree(int n) : n(n), tree(n + 1, 0) {}

  // Build from array — O(N log N) simple, O(N) optimized below
  FenwickTree(int n, vector<int> &a) : n(n), tree(n + 1, 0) {
    for (int i = 0; i < n; i++)
      update(i, a[i]);
  }

  // Add delta to index i (0-indexed)
  void update(int i, long long delta) {
    for (++i; i <= n; i += i & (-i))
      tree[i] += delta;
  }

  // Prefix sum [0..i] (0-indexed)
  long long query(int i) {
    long long sum = 0;
    for (++i; i > 0; i -= i & (-i))
      sum += tree[i];
    return sum;
  }

  // Range sum [l..r] (0-indexed)
  long long query(int l, int r) {
    return query(r) - (l > 0 ? query(l - 1) : 0);
  }

  // Set a[i] = val (requires knowing old value)
  void set(int i, long long old_val, long long new_val) {
    update(i, new_val - old_val);
  }
};

// Fenwick 2D — for 2D range sum queries
struct Fenwick2D {
  int n, m;
  vector<vector<long long>> tree;

  Fenwick2D(int n, int m)
      : n(n), m(m), tree(n + 1, vector<long long>(m + 1, 0)) {}

  void update(int x, int y, long long delta) {
    for (int i = x + 1; i <= n; i += i & (-i))
      for (int j = y + 1; j <= m; j += j & (-j))
        tree[i][j] += delta;
  }

  long long query(int x, int y) {
    long long sum = 0;
    for (int i = x + 1; i > 0; i -= i & (-i))
      for (int j = y + 1; j > 0; j -= j & (-j))
        sum += tree[i][j];
    return sum;
  }

  // Range sum [x1..x2][y1..y2]
  long long query(int x1, int y1, int x2, int y2) {
    return query(x2, y2) - query(x1 - 1, y2) - query(x2, y1 - 1) +
           query(x1 - 1, y1 - 1);
  }
};

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int n, q;
  cin >> n >> q;
  vector<int> a(n);
  for (auto &x : a)
    cin >> x;

  FenwickTree ft(n, a);

  while (q--) {
    int type;
    cin >> type;
    if (type == 1) {
      int i;
      long long val;
      cin >> i >> val;
      ft.update(i - 1, val); // 1-indexed
    } else {
      int l, r;
      cin >> l >> r;
      cout << ft.query(l - 1, r - 1) << "\n"; // 1-indexed
    }
  }
  return 0;
}
